test/support/utils_sycl.h

```diff
@@ -139,28 +139,88 @@ make_new_policy(_Policy&& __policy)
 // create the queue with custom asynchronous exceptions handler
 static auto my_queue = sycl::queue(default_selector, async_handler);
 
-// Invoke op(policy,rest...) for each possible policy.
+// check fp16/fp64 support by a device
+template<typename T>
+bool has_type_support(const sycl::device&) { return true; }
+
+template<>
+inline bool has_type_support<double>(const sycl::device& device)
+{
+    return device.has(sycl::aspect::fp64);
+}
+
+template<>
+inline bool has_type_support<sycl::half>(const sycl::device& device)
+{
+    return device.has(sycl::aspect::fp16);
+}
+
+template <typename, typename = void>
+struct has_value_type : ::std::false_type {};
+
+template <typename T>
+struct has_value_type<T, ::std::void_t<typename std::iterator_traits<T>::value_type>> : ::std::true_type {};
+
+template<typename T>
+std::enable_if_t<!has_value_type<T>::value, bool> has_iter_type_support(const sycl::device& device)
+{
+    return has_type_support<T>(device);
+};
+
+template<typename T>
+std::enable_if_t<has_value_type<T>::value, bool> has_iter_type_support(const sycl::device& device)
+{
+    return has_type_support<typename std::iterator_traits<T>::value_type>(device);
```

Comment:
Additionally, am argument can be a simple value type (init value for example), not std::iterator_traits<T>::value_type...

Reply:
Discussed offline and decided to leave it as is, since we need to distinguish iterators and other types.